<a href="https://colab.research.google.com/github/artificialrick79/Studiare-AI/blob/Machine-Learning-con-python/ML_Train_Test_split_con_scikit_learn.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Importiamo pandas e numpy e poi un dataset di nome boston disponibile su scikit learn

In [1]:
import pandas as pd
import numpy as np

In [3]:
from sklearn.datasets import load_boston

In [4]:
boston = load_boston()

I dati sono stati caricati in una speciale struttura dati di scikit learn molto simile ad un dizionario. Se vogliamo ottenere le informazioni relative al dataset possiamo usare la funzione DESCR


In [8]:
boston.DESCR

".. _boston_dataset:\n\nBoston house prices dataset\n---------------------------\n\n**Data Set Characteristics:**  \n\n    :Number of Instances: 506 \n\n    :Number of Attributes: 13 numeric/categorical predictive. Median Value (attribute 14) is usually the target.\n\n    :Attribute Information (in order):\n        - CRIM     per capita crime rate by town\n        - ZN       proportion of residential land zoned for lots over 25,000 sq.ft.\n        - INDUS    proportion of non-retail business acres per town\n        - CHAS     Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)\n        - NOX      nitric oxides concentration (parts per 10 million)\n        - RM       average number of rooms per dwelling\n        - AGE      proportion of owner-occupied units built prior to 1940\n        - DIS      weighted distances to five Boston employment centres\n        - RAD      index of accessibility to radial highways\n        - TAX      full-value property-tax rate per $10,000

Noi per questioni di chiarezza importiamo il dataset in array numpy utilizzando l'attributo "data" per ottenere le features e "target" per ottenere i target

In [9]:
X = boston.data
Y = boston.target

Per decidere come ripartire il dataset vediamo prima quanto è grande. Fino a 10k esempi la ripartizione è 70%-30% per poi passare a 98%-2% per dataset oltre il milione.

In [11]:
X.shape

(506, 13)

Il dataset in questione ha solo 506 esempi e quindi la porzione dedicata al test sarà il 30%.

Per eseguire una suddivisione del dataset utilizziamo una funzione già disponibile in scikit learn che si chiama "train test split"

In [10]:
from sklearn.model_selection import train_test_split

In [12]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.3)

Vediamo ora se è stato effettivamente ripartito tra train e test

In [14]:
X_train.shape

(354, 13)

In [15]:
X_test.shape

(152, 13)

Per effettuare la suddivisione di un dataframe non c'è una funzione definita ma dobbiamo inventarcela noi ma comunque è un'attività che non capiterà mai di fare.

In [17]:
boston_df = pd.DataFrame(data= np.c_[boston['data'], boston['target']], columns= np.append(boston['feature_names'], 'TARGET'))

Andiamo a vedere se il dataframe è stato creato.

In [18]:
boston_df.head()

CRIM    ZN  INDUS  CHAS    NOX  ...    TAX  PTRATIO       B  LSTAT  TARGET
0  0.00632  18.0   2.31   0.0  0.538  ...  296.0     15.3  396.90   4.98    24.0
1  0.02731   0.0   7.07   0.0  0.469  ...  242.0     17.8  396.90   9.14    21.6
2  0.02729   0.0   7.07   0.0  0.469  ...  242.0     17.8  392.83   4.03    34.7
3  0.03237   0.0   2.18   0.0  0.458  ...  222.0     18.7  394.63   2.94    33.4
4  0.06905   0.0   2.18   0.0  0.458  ...  222.0     18.7  396.90   5.33    36.2

[5 rows x 14 columns]

Utilizziamo la funzione "sample" per creare un test set a partire dal dataset (in questo caso il 30%).

In [20]:
boston_test_df = boston_df.sample(frac=0.3)

Per ottenere il train set dobbiamo togliere tutti gli elementi presenti nel test set a partire dai suoi indici

In [21]:
boston_train_df = boston_df.drop(boston_test_df.index)

Controlliamo se la suddivisione tra train set e test set è stata fatta correttamente

In [22]:
boston_test_df.shape[0]

152

In [23]:
boston_train_df.shape[0]

354